lib/core/src/irods_service_account.cpp

```diff
@@ -0,0 +1,132 @@
+
+        std::string svc_acct_cfg_fn;
+        irods::error ret = irods::get_full_path_for_config_file(SERVICE_ACCOUNT_CONFIG_FILE, svc_acct_cfg_fn);
+
+        if (!ret.ok()) {
+            // if config file not found, the machine probably does not have a server
+            return false;
+        }
+
+        std::ifstream svc_acct_cfg_strm{svc_acct_cfg_fn};
+        if (!svc_acct_cfg_strm.is_open()) {
+            // if we can't open the config file, it's likely due to a permission issue
+            // meaning it is extremely unlikely we are the service account
+            return false;
+        }
+
+        kvp_map_t service_acct_cfg;
+        uint_fast64_t cfg_line_no = 0;
+        for (std::string cfg_line; std::getline(svc_acct_cfg_strm, cfg_line);) {
+            cfg_line_no++;
+            boost::algorithm::trim(cfg_line); // trim whitespace (should handle CRLF)
+            if (cfg_line.empty()) {
+                // ignore empty lines
+                continue;
+            }
+            if (cfg_line.at(0) == '#') {
+                // ignore lines starting with #
+                continue;
+            }
+
+            // For now, let's just re-use our kvp-parsing functionality
+            // FIXME: does not handle spaces around "=", quotes, inline comments, or escape-newline
+            ret = parse_escaped_kvp_string(cfg_line, service_acct_cfg);
+            if (!ret.ok()) {
+                irods::log(LOG_WARNING,
+                           fmt::format("{} - {} parse failure on line {}",
+                                       __PRETTY_FUNCTION__,
+                                       SERVICE_ACCOUNT_CONFIG_FILE,
+                                       cfg_line_no));
+            }
+        }
+
+        const auto& svc_acct_itr = service_acct_cfg.find(SERTICE_ACCOUNT_KW);
+        if (svc_acct_itr == service_acct_cfg.end()) {
+            THROW(
+                KEY_NOT_FOUND, fmt::format("could not find {} in {}", SERTICE_ACCOUNT_KW, SERVICE_ACCOUNT_CONFIG_FILE));
+        }
+        service_account_username = svc_acct_itr->second;
+
+        // compare uids rather than usernames, to account for overlapping uids
+        // NOTE: do not reference the strings inside these structs, as their buffer goes out of scope
+        struct passwd svc_acct_pwd{};
+        struct passwd* svc_acct_pwdp = nullptr;
+        const auto _pwd_init_len = sysconf(_SC_GETPW_R_SIZE_MAX);
+        size_t pwd_buf_len = PWD_BUF_LEN_DEFAULT; // a reasonable buffer size in case sysconf gives us something awful
+        if (_pwd_init_len >= PWD_BUF_LEN_MINIMUM) {
+            pwd_buf_len = static_cast<size_t>(_pwd_init_len);
+        }
+        int pwd_e = 0;
+        while (true) {
+            // this sure is a lot of fuss to allocate a buffer for data we're not going to use anyway...
+            const std::unique_ptr<char[]> pwd_buf( // NOLINT(cppcoreguidelines-avoid-c-arrays, modernize-avoid-c-arrays)
+                new char[pwd_buf_len]);
```

Comment:
Does the following work?
```cpp
std::array<char, pwd_buf_len> pwd_buf;
pwd_e = getpwnam_r(..., pwd_buf.data());
```

Reply:
Unfortunately, no, already tried that

Reply:
I'm guessing it's because the buffer size isn't known at compile-time.

Carry on.